File: CSE216/HW3/square.py
from rectangle import Rectangle
from quadrilateral import Quadrilateral


class Square(Rectangle):

    def __init__(self, *floats):
        super().__init__(*floats)
        if not self.__is_member():
            raise TypeError("A square cannot be formed by the given coordinates.")

    def __eq__(self, other):
        return isinstance(other, Square) and super().__eq__(other)

    def __str__(self):
        string = "Square: "
        for pt in self.vertices:
            string += pt.__str__() + ", "
        return string[0:len(string) - 2]

    # Ask Banerjee about this!!!
    # if super().__is_member():
    #     sides = self.side_lengths()
    #     if sides[0] == sides[1]:
    #         return True
    #     else:
    #         return False
    def __is_member(self):
        sides = self.side_lengths()
        return sides[0] == sides[1] and sides[1] == sides[2] and sides[2] == sides[3] and sides[3] == sides[0]

    def get_is_member(self):
        return self.__is_member()

    def snap(self):
        """Snaps the sides of the square such that each corner (x,y) is modified to be a corner (x',y') where x' is the
        integer value closest to x and y' is the integer value closest to y. This, of course, may change the shape to a
        general quadrilateral, hence the return type. The only exception is when the square is positioned in a way where
        this approximation will lead it to vanish into a single point. In that case, a call to snap() will not modify
        this square in any way."""
        new_square = Quadrilateral(round(self.vertices[0].x), round(self.vertices[0].y), round(self.vertices[1].x),
                                   round(self.vertices[1].y), round(self.vertices[2].x), round(self.vertices[2].y),
                                   round(self.vertices[3].x), round(self.vertices[3].y))

        for i in range(3):
            j = 1 + i
            for j in range(i):
                if new_square.vertices[i].x == new_square.vertices[j].x and new_square.vertices[i].y == \
                        new_square.vertices[j].y:
                    return self

        return new_square

File: CSE216/HW2/Square.java
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.DoubleStream;

public class Square extends Rectangle implements Snappable {

    // TODO: this constructor must NOT be changed. Instead, resolve the error by adding code elsewhere.
    public Square(List<TwoDPoint> vertices) {
        super(vertices);
    }

    /**
     * Given a list of vertices assumed to be provided in a counterclockwise order in a two-dimensional plane, checks
     * whether or not they constitute a valid square.
     *
     * @param vertices the specified list of vertices in a counterclockwise order
     * @return <code>true</code> if the four vertices can form a square, <code>false</code> otherwise.aw
     *
     */
    @Override
    public boolean isMember(List<? extends Point> vertices) {
        return vertices.size() == 4 &&
                DoubleStream.of(getSideLengths()).boxed().collect(Collectors.toSet()).size() == 1;
    }

    /**
     * Snaps the sides of the square such that each corner (x,y) is modified to be a corner (x',y') where x' is the
     * the integer value closest to x and y' is the integer value closest to y. This, of course, may change the shape
     * to a general quadrilateral, hence the return type. The only exception is when the square is positioned in a way
     * where this approximation will lead it to vanish into a single point. In that case, a call to {@link #snap()}
     * will not modify this square in any way.
     */
    @Override
    public Quadrilateral snap() {
        // TODO
        double[] values = new double[8];
        int j = 0;
        for(int i = 0; i < values.length; i = i + 2){
            values[i] = Math.round(getVertices()[j].getX());
            values[i+1] = Math.round(getVertices()[j++].getY());
        }

        List<TwoDPoint> newVertices = TwoDPoint.ofDoubles(values);
        for(int i = 0; i < newVertices.size(); i++){
            for(j = 1 + i; j < newVertices.size(); j++){
                if(newVertices.get(i).getX() == newVertices.get(j).getX() && newVertices.get(i).getY() == newVertices.get(j).getY()){
                    return this;
                }
            }
        }
        return new Quadrilateral(newVertices);
    }
}
